### DISCUSSION TITLE
Newer versions of PyInstaller / Python make rdiff-backup don't find their modules

### DISCUSSION THREAD

**Discussion by ericzolf** [pyinstaller/pyinstaller#7078] [Help] [answered]
Basically, I could use some help on rdiff-backup/rdiff-backup#731

## Gory details

1. building rdiff-backup.exe works with PyInstaller 5.1 and Python 3.10.5 (resp. 5.0 and 3.10.4)
2. ISSUE A: building with 5.2/5.3 and 3.10.6/3.10.7 leads to a binary which doesn't find its modules:

```
> rdiff-backup.exe --version
Traceback (most recent call last):
  File "rdiff-backup", line 19, in <module>
ModuleNotFoundError: No module named 'rdiffbackup'
[1156] Failed to execute script 'rdiff-backup' due to unhandled exception!
```

3. ISSUE B: building with 5.2/5.3 and 3.10.4 (and probably 3.10.5 but I didn't test) leads to a binary which doesn't find the submodules collected via hooks (basically, `rdiff-backup.exe --help` shows an empty list of actions):

```
$ head -n-0 tools/hook-rdiffbackup.*
==> tools/hook-rdiffbackup.actions_mgr.py <==
from PyInstaller.utils.hooks import collect_submodules
hiddenimports = collect_submodules('rdiffbackup.actions')

==> tools/hook-rdiffbackup.meta_mgr.py <==
from PyInstaller.utils.hooks import collect_submodules
hiddenimports = collect_submodules('rdiffbackup.meta')
```

> **NOTE:** I didn't fully analyze ISSUE B because I'm stuck first with ISSUE A...

4. using the development version `pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip` leads to the same issue as in point 2

## My questions

1. going through the release notes of PyInstaller, I could imagine that ISSUE B is due to the Bugfixes #6846 and/or #6850 but I don't understand what should be wrong and how to address
2. I have no clue where to look for ISSUE A, seems to me an incompatibility of PyInstaller with Python 3.10.6/3.10.7

## Code context

PyInstaller is called [as part of the pipeline](https://github.com/rdiff-backup/rdiff-backup/blob/master/.github/workflows/deploy.yml#L96) resp. in an [Ansible playbook for my Vagrant VM](https://github.com/rdiff-backup/rdiff-backup/blob/master/tools/windows/playbook-pipeline.yml#L26), and they both call PyInstaller through the [tools/win_build_rdiffbackup.sh](https://github.com/rdiff-backup/rdiff-backup/blob/master/tools/win_build_rdiffbackup.sh#L37-L43) script.

For issue B, the submodules/plugins are loaded by the [action manager](https://github.com/rdiff-backup/rdiff-backup/blob/master/src/rdiffbackup/actions_mgr.py) or the [meta manager](https://github.com/rdiff-backup/rdiff-backup/blob/master/src/rdiffbackup/meta_mgr.py) (and as written, it works with older versions of PyInstaller/Python).

Let me know if you need more information.

**Comment by rokm** [comment, marked as answer]
When [a direct import from your frozen entry-point script](https://github.com/rdiff-backup/rdiff-backup/blob/f8e4dd58af098a1812342e7b1df1deefdd939ea5/src/rdiff-backup#L19) cannot be found at runtime, this usually means that you have messed up your environment and the module/package is not discoverable during the PyInstaller run (usually that means that the module's/package's parent location is not in `sys.path` during the build).

Looking at [your build script](https://github.com/rdiff-backup/rdiff-backup/blob/f8e4dd58af098a1812342e7b1df1deefdd939ea5/tools/win_build_rdiffbackup.sh#L38-L40), I see you are using `--paths` to pass extra search paths;
```
	--paths=build/lib.${py_win_bits}-${py_ver_brief} \
        --paths=${LIBRSYNC_DIR}/lib \
        --paths=${LIBRSYNC_DIR}/bin \
```
which, during runtime (and with added directory where the entry-point script resides), translate into:
```
['D:\\a\\rdiff-backup\\rdiff-backup\\build\\scripts-3.10',
 'D:\\a\\rdiff-backup\\rdiff-backup\\build\\lib.win-amd64-3.10',
 'C:\\Users\\runneradmin\\librsync64\\lib',
 'C:\\Users\\runneradmin\\librsync64\\bin']
``` 
So, where does the `D:\\a\\rdiff-backup\\rdiff-backup\\build\\lib.win-amd64-3.10` directory come from? As far as I can tell, you are (ab)using the [`python setup.py build_distwheel` command](https://github.com/rdiff-backup/rdiff-backup/blob/f8e4dd58af098a1812342e7b1df1deefdd939ea5/tools/win_build_rdiffbackup.sh#L35) to create it for you.

But the name of the directory depends on python version (or more likely, its bundled `pip` and/or `setuptools`):
* [here](https://github.com/rdiff-backup/rdiff-backup/runs/8284065271?check_suite_focus=true#step:6:21) is what you get during the run with python 3.10.5: ```creating build\lib.win-amd64-3.10```
* and [here](https://github.com/rdiff-backup/rdiff-backup/runs/8283608353?check_suite_focus=true#step:6:21) is what you get during the run with python 3.10.7: ```creating build\lib.win-amd64-cpython-310```

So unless you have a very good reason not to, I suggest you install the built wheel into your python environment, so that `rdiffbackup` is available without manually mucking around with search paths. This will also allow you to collect the package's metadata using `--copy-medatata rdiff-backup` (or `copy_metadata('rdiff-backup')` in a hook) instead of [manually copying the directory via --add-data](https://github.com/rdiff-backup/rdiff-backup/blob/f8e4dd58af098a1812342e7b1df1deefdd939ea5/tools/win_build_rdiffbackup.sh#L43).

**Reply by ericzolf** [reply]
OK, I got it, the change of path in a minor version is a bummer. I don't know if it's a good reason, but I don't like to install development artifacts on the system. It's not an issue in a pipeline but not optimal on a development workstation. Anyway, that gives me enough to chew on.

**Reply by rokm** [reply]
> I don't know if it's a good reason, but I don't like to install development artifacts on the system. It's not an issue in a pipeline but not optimal on a development workstation.

In that case, create a venv, install everything into it, and discard it once the build is complete.

**Reply by ericzolf** [reply]
Good idea, sounds even like a plan for tox. Let me check and report back.

**Reply by ericzolf** [reply]
As a workaround for now, fixing the path solved issue A, but issue B is still there as soon as I upgrade to PyInstaller 5.2 or 5.3. It can wait until I've got a cleaner approach in place.

**Reply by rokm** [reply]
Hmm, well, I tried installing win64 `rdist-backup` wheel from `package-artifacts.zip` from your latest `Deploy-All` run (and building `rsync.dll` from source and making it available  - shouldn't it be included in the wheel?). Then, I tried freezing the following python script 
```python
# program.py
from rdiffbackup import actions_mgr
print(actions_mgr.get_actions_dict())
```
using PyInstaller 5.3 (using your two hooks from the `tools`): 
```pyinstaller --clean --noconfirm program.py --additional-hooks-dir "rdiff-backup-src\tools" --onefile```

The resulting executable produces the following output:
```
backup <class 'rdiffbackup.actions.backup.BackupAction'>
calculate <class 'rdiffbackup.actions.calculate.CalculateAction'>
remove <class 'rdiffbackup.actions.remove.RemoveAction'>
compare <class 'rdiffbackup.actions.compare.CompareAction'>
server <class 'rdiffbackup.actions.server.ServerAction'>
regress <class 'rdiffbackup.actions.regress.RegressAction'>
restore <class 'rdiffbackup.actions.restore.RestoreAction'>
test <class 'rdiffbackup.actions.test.TestAction'>
info <class 'rdiffbackup.actions.info.InfoAction'>
list <class 'rdiffbackup.actions.list_.ListAction'>
verify <class 'rdiffbackup.actions.verify.VerifyAction'>
```

So the `rdiffbackup.actions` submodules are properly collected and listed back at runtime. The `rdiffbackup.meta` ones are not, though, and the hook emits the following warnings:
```
328 WARNING: Failed to collect submodules for 'rdiffbackup.meta' because importing 'rdiffbackup.meta' raised: AttributeError: partially initialized module 'rdiffbackup.meta' has no attribute 'FlatExtractor' (most likely due to a circular import)
```
(also [seen](https://github.com/rdiff-backup/rdiff-backup/runs/8286713823?check_suite_focus=true#step:6:394) in the `Deploy-All` run).

And the module is in fact not importable, even in unfrozen version:
```
>python
Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import  rdiffbackup.meta
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "[...]\venv\lib\site-packages\rdiffbackup\meta\__init__.py", line 23, in <module>
    from rdiff_backup import log, rpath
  File "[...]\venv\lib\site-packages\rdiff_backup\rpath.py", line 48, in <module>
    from rdiffbackup.meta import acl_posix, acl_win, ea
  File "[...]\pyi-rdiff\venv\lib\site-packages\rdiffbackup\meta\acl_posix.py", line 39, in <module>
    from . import ea
  File "[...]\venv\lib\site-packages\rdiffbackup\meta\ea.py", line 168, in <module>
    class EAExtractor(meta.FlatExtractor):
AttributeError: partially initialized module 'rdiffbackup.meta' has no attribute 'FlatExtractor' (most likely due to a circular import)
```

**Reply by ericzolf** [reply]
* on the wheel / rsync.dll topic - you're most probably right, nobody complained so far and I'm not a big fan of Windows, so probably nobody uses the wheel under Windows... I've added rdiff-backup/rdiff-backup#733
* interestingly enough I couldn't reproduce the meta-issue at first, because I was first loading rdiffbackup.actions, but if I load first meta, then I get the same error as you:

```
C:\Users\vagrant\Develop\rdiff-backup>.tox\py\Scripts\ac
tivate.bat

C:\Users\vagrant\Develop\rdiff-backup>where python 
C:\Users\vagrant\Develop\rdiff-backup\.tox\py\Scripts\python.exe
C:\Python64\python.exe

C:\Users\vagrant\Develop\rdiff-backup>python       
Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (
AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import rdiffbackup.actions 
>>> import rdiffbackup.meta 
>>>  exit()

C:\Users\vagrant\Develop\rdiff-backup>python       
Python 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (
AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import rdiffbackup.meta 
Traceback (most recent call last): 
  File "<stdin>", line 1, in <module> 
  File "C:\Users\vagrant\Develop\rdiff-backup\.tox\py\lib\site-packages\rdiffbac
kup\meta\__init__.py", line 23, in <module>
    from rdiff_backup import log, rpath
  File "C:\Users\vagrant\Develop\rdiff-backup\.tox\py\lib\site-packages\rdiff_ba
ckup\rpath.py", line 48, in <module>
    from rdiffbackup.meta import acl_posix, acl_win, ea
  File "C:\Users\vagrant\Develop\rdiff-backup\.tox\py\lib\site-packages\rdiffbac
kup\meta\acl_posix.py", line 39, in <module>
    from . import ea
  File "C:\Users\vagrant\Develop\rdiff-backup\.tox\py\lib\site-packages\rdiffbac
kup\meta\ea.py", line 168, in <module>
    class EAExtractor(meta.FlatExtractor):
AttributeError: partially initialized module 'rdiffbackup.meta' has no attribute
 'FlatExtractor' (most likely due to a circular import)
```

OK, so I need to break the circular dependency meta -> rpath -> acl_posix/ea -> meta for PyInstaller (now 5.4) to be happy. Probably not a bad thing to do anyway, but I'm a bit disturbed by the fact that I only need to do this for PyInstaller, where it didn't have an issue with this till 5.1.

Thanks for all the help so far, I know at least where to look further, and I have got a better release process using tox as well.
